frame/atomic-swap/src/lib.rs

```diff
@@ -0,0 +1,230 @@
+	/// The overarching event type.
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+	/// The currency mechanism.
+	type Currency: ReservableCurrency<Self::AccountId>;
+	/// Limit of proof size.
+	type ProofLimit: Get<u32>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as AtomicSwap {
+		pub PendingSwaps: double_map
+			hasher(twox_64_concat) T::AccountId, hasher(blake2_128_concat) HashedProof
+			=> Option<PendingSwapFor<T>>;
+	}
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// Swap already exists.
+		AlreadyExist,
+		/// Swap proof is invalid.
+		InvalidProof,
+		/// Proof is too large.
+		ProofTooLarge,
+		/// Source does not match.
+		SourceMismatch,
+		/// Swap does not exist.
+		NotExist,
+		/// Duration has not yet passed for the swap to be cancelled.
+		DurationNotPassed,
+	}
+}
+
+decl_event!(
+	/// Event of atomic swap pallet.
+	pub enum Event<T> where
+		Balance = BalanceFor<T>,
+		AccountId = AccountIdFor<T>,
+		PendingSwap = PendingSwapFor<T>,
+	{
+		/// Swap created.
+		NewSwap(AccountId, HashedProof, PendingSwap),
+		/// Swap claimed.
+		SwapClaimed(AccountId, HashedProof, Balance),
+		/// Swap cancelled.
+		SwapCancelled(AccountId, HashedProof),
+	}
+);
+
+decl_module! {
+	/// Module definition of atomic swap pallet.
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+		/// Register a new atomic swap, declaring an intention to send funds from origin to target
+		/// on the current blockchain. The target can claim the fund using the revealed proof. If
+		/// the fund is not claimed after `duration` blocks, then the sender can cancel the swap.
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// - `target`: Receiver of the atomic swap.
+		/// - `hashed_proof`: The blake2_256 hash of the secret proof.
+		/// - `balance`: Funds to be sent from origin.
+		/// - `duration`: Locked duration of the atomic swap.
+		#[weight = T::DbWeight::get().reads_writes(1, 1).saturating_add(40_000_000)]
+		fn create_swap(
+			origin,
+			target: AccountIdFor<T>,
+			hashed_proof: HashedProof,
+			balance: BalanceFor<T>,
+			duration: BlockNumberFor<T>,
+		) {
+			let source = ensure_signed(origin)?;
+			ensure!(
+				PendingSwaps::<T>::get(&target, hashed_proof).is_none(),
+				Error::<T>::AlreadyExist
+			);
+
+			T::Currency::reserve(&source, balance)?;
+
+			let swap = PendingSwap {
+				source,
+				balance,
+				end_block: frame_system::Module::<T>::block_number() + duration,
+			};
+			PendingSwaps::<T>::insert(target.clone(), hashed_proof.clone(), swap.clone());
+
+			Self::deposit_event(
+				RawEvent::NewSwap(target, hashed_proof, swap)
+			);
+		}
+
+		/// Claim an atomic swap.
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// - `proof`: Revealed proof of the claim.
+		#[weight = T::DbWeight::get().reads_writes(1, 1)
```

Comment:
```suggestion
		#[weight = T::DbWeight::get().reads_writes(2, 2)
```

There will be one more read/write here because we will be modifying the balance of an account that is not the caller. (the swap source)